File: src/Java/algorithms/linkedlist/SwapNodesInPairs.java
package algorithms.linkedlist; 

import java.util.*; 
import lib.ListNode; 

public class SwapNodesInPairs {
    /* Function: Swap each two nodes in a singly linked list. 
     * Solution: Two pointers is enough. 
     */ 
    
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) {
            return head; 
        }
        
        ListNode dummy = new ListNode(0); 
        dummy.next = head; 
        
        ListNode pre = dummy, cur = head; 
        while (cur != null && cur.next != null) { // Here is and not or!
            pre.next = cur.next; 
            cur.next = cur.next.next; 
            pre.next.next = cur; 
            
            pre = cur; 
            cur = cur.next; 
        }
        
        return dummy.next; 
    }
}

File: src/CPP/src/leetcode/SwapNodesInPairs.cpp
/*
 * [Source] https://leetcode.com/problems/swap-nodes-in-pairs/
 * [Difficulty]: Easy
 * [Tag]: Linked List
 */

#include <iostream>

using namespace std;

// [Solution]: pre->a->b, so next=b->next, b->next=a, a->next=next, pre->next=b, pre=next
// [Corner Case]: If there are odd number of nodes
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/linkedlist/SwapNodesInPairs.java
 */

/* Java solution
public class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) {
            return head; 
        }
        ListNode slow = head; 
        ListNode fast = head.next; 
        ListNode cur = new ListNode(0); 
        head = head.next; 
        while (fast != null) {
            cur.next = fast; 
            slow.next = fast.next; 
            fast.next = slow; 
            cur = slow; 
            slow = slow.next; 
            fast = (slow == null) ? null : slow.next; 
        }
        return head; 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
